server/indexer.go

```diff
@@ -0,0 +1,231 @@
+func (kv *KVIndexer) IndexBlock(block *tmtypes.Block, txResults []*abci.ResponseDeliverTx) error {
+	height := block.Header.Height
+
+	batch := kv.db.NewBatch()
+	defer batch.Close()
+
+	// record index of valid eth tx during the iteration
+	var ethTxIndex int32
+	for txIndex, tx := range block.Txs {
+		result := txResults[txIndex]
+		if !backend.TxSuccessOrExceedsBlockGasLimit(result) {
+			continue
+		}
+
+		tx, err := kv.clientCtx.TxConfig.TxDecoder()(tx)
+		if err != nil {
+			kv.logger.Error("Fail to decode tx", "err", err, "block", height, "txIndex", txIndex)
+			continue
+		}
+
+		if !isEthTx(tx) {
+			continue
+		}
+
+		txs, err := rpctypes.ParseTxResult(result, tx)
+		if err != nil {
+			kv.logger.Error("Fail to parse event", "err", err, "block", height, "txIndex", txIndex)
+			continue
+		}
+
+		var cumulativeGasUsed uint64
+		for msgIndex, msg := range tx.GetMsgs() {
+			ethMsg := msg.(*evmtypes.MsgEthereumTx)
+			txHash := common.HexToHash(ethMsg.Hash)
+
+			txResult := ethermint.TxResult{
+				Height:     height,
+				TxIndex:    uint32(txIndex),
+				MsgIndex:   uint32(msgIndex),
+				EthTxIndex: ethTxIndex,
+			}
+			if result.Code != abci.CodeTypeOK {
+				// exceeds block gas limit scenario, set gas used to gas limit because that's what's charged by ante handler.
+				// some old versions don't emit any events, so workaround here directly.
+				txResult.GasUsed = ethMsg.GetGas()
+				txResult.Failed = true
+			} else {
+				parsedTx := txs.GetTxByMsgIndex(msgIndex)
+				if parsedTx == nil {
+					kv.logger.Error("msg index not found in events", "msgIndex", msgIndex)
+					continue
+				}
+				if parsedTx.EthTxIndex >= 0 && parsedTx.EthTxIndex != ethTxIndex {
+					kv.logger.Error("eth tx index don't match", "expect", ethTxIndex, "found", parsedTx.EthTxIndex)
+				}
+				txResult.GasUsed = parsedTx.GasUsed
+				txResult.Failed = parsedTx.Failed
+			}
+
+			cumulativeGasUsed += txResult.GasUsed
+			txResult.CumulativeGasUsed = cumulativeGasUsed
+			ethTxIndex++
+
+			if err := saveTxResult(kv.clientCtx.Codec, batch, txHash, &txResult); err != nil {
+				return sdkerrors.Wrapf(err, "IndexBlock %d", height)
+			}
+		}
+	}
+	if err := batch.Write(); err != nil {
+		return sdkerrors.Wrapf(err, "IndexBlock %d, write batch", block.Height)
+	}
+	return nil
+}
+
+// LastIndexedBlock returns the latest indexed block number, returns -1 if db is empty
+func (kv *KVIndexer) LastIndexedBlock() (int64, error) {
+	return LoadLastBlock(kv.db)
+}
+
+// FirstIndexedBlock returns the first indexed block number, returns -1 if db is empty
+func (kv *KVIndexer) FirstIndexedBlock() (int64, error) {
+	return LoadFirstBlock(kv.db)
+}
+
+// isEthTx check if the tx is an eth tx
+func isEthTx(tx sdk.Tx) bool {
+	extTx, ok := tx.(authante.HasExtensionOptionsTx)
+	if !ok {
+		return false
+	}
+	opts := extTx.GetExtensionOptions()
+	if len(opts) != 1 || opts[0].GetTypeUrl() != "/ethermint.evm.v1.ExtensionOptionsEthereumTx" {
+		return false
+	}
+	return true
+}
```

Comment:
move this to the bottom of the file or to `util.go`

Reply:
done